File: ios/RecipeClipper/App/AutoBackup.swift
import Combine
import Foundation

/// Where the automatic copy (#150) goes (Android's `BackupFolder`): on iOS the app's own iCloud
/// Drive folder. Every call may touch the file system or iCloud, so none runs on the main thread.
protocol BackupFolder: AnyObject {
    /// `.ready` while there is somewhere to write, `.unavailable` otherwise.
    func destination() async -> BackupDestination
    /// The names of the files in the folder, or nil if it can't be read.
    func list() async -> [String]?
    /// Writes a zip export (`BackupArchive`) as `name`, replacing a file of that name; the name
    /// it ended up with, or nil if it couldn't. A failed write leaves no partial file.
    func write(name: String, json: String, photos: [String: URL]) async -> String?
    /// Deletes the named file, if it is there.
    func delete(name: String) async
}

/// Where `AutoBackupRecord` is kept (Android's `AutoBackupStore`).
protocol AutoBackupStore: AnyObject {
    var record: AutoBackupRecord { get }
    func update(_ transform: (inout AutoBackupRecord) -> Void)
}

/// What Settings shows: the record, and whether its destination can be written now.
struct AutoBackupState: Equatable {
    var record: AutoBackupRecord
    var destination: BackupDestination
}

enum AutoBackupOutcome: Equatable {
    case written, upToDate, off, noDestination, failed
}

/// The automatic copy (#150; Android's `AutoBackup`): the latest export, photos included, as a
/// zip in the app's iCloud Drive folder ("Recipe Clipper" in Files), keeping the newest three.
/// The rules are `AutoBackupPolicy`'s. Nothing is written anywhere else and nothing is sent to
/// any server: iCloud Drive syncs the folder as it does any document.
@MainActor
final class AutoBackup {
    private let backups: BackupRepository
    private let folder: BackupFolder
    private let store: AutoBackupStore
    private let clock: Clock
    private let subject: CurrentValueSubject<AutoBackupState, Never>
    private var inFlight: Task<AutoBackupOutcome, Never>?

    init(backups: BackupRepository, folder: BackupFolder, store: AutoBackupStore, clock: Clock) {
        self.backups = backups
        self.folder = folder
        self.store = store
        self.clock = clock
        // Unknown until asked (asking may block): shown as unavailable only once iCloud says so.
        subject = CurrentValueSubject(AutoBackupState(record: store.record, destination: .ready))
    }

    /// The record and destination, now and on every change.
    var state: AnyPublisher<AutoBackupState, Never> { subject.eraseToAnyPublisher() }
    var current: AutoBackupState { subject.value }

    /// Asks iCloud again, e.g. when Settings opens. Returns the work so a test can await it.
    @discardableResult
    func refreshDestination() -> Task<Void, Never> {
        Task { [folder] in
            let destination = await folder.destination()
            self.publish(destination: destination)
        }
    }

    func setEnabled(_ enabled: Bool) {
        store.update { $0.enabled = enabled }
        publish()
    }

    /// Writes a copy if `AutoBackupPolicy.isDue` (always for `force`: "Back up now", which works
    /// even with the automatic copy off). Old copies beyond the newest three go afterwards, so
    /// a failed write never leaves fewer. One at a time: a second call joins the first.
    @discardableResult
    func run(force: Bool = false) async -> AutoBackupOutcome {
        if let inFlight { return await inFlight.value }
        let task = Task { await perform(force: force) }
        inFlight = task
        let outcome = await task.value
        inFlight = nil
        return outcome
    }

    private func perform(force: Bool) async -> AutoBackupOutcome {
        let record = store.record
        if !force && !record.enabled { return .off }
        let destination = await folder.destination()
        publish(destination: destination)
        guard destination == .ready else { return .noDestination }
        guard let existing = await folder.list() else { return failed() }
        guard case .success(let exported) = await backups.export() else { return failed() }
        let now = clock.now()
        let fingerprint = AutoBackupPolicy.fingerprint(json: exported.json, photos: Array(exported.photos.keys))
        guard AutoBackupPolicy.isDue(record, fingerprint: fingerprint, existing: existing, now: now, force: force) else {
            return .upToDate
        }
        guard let written = await folder.write(
            name: AutoBackupPolicy.fileName(now: now), json: exported.json, photos: exported.photos
        ) else { return failed() }
        for name in AutoBackupPolicy.toDelete(existing: existing + [written], written: written) {
            await folder.delete(name: name)
        }
        store.update {
            $0.lastBackupAt = now
            $0.lastFingerprint = fingerprint
            $0.lastFile = written
            $0.lastFailed = false
        }
        publish()
        return .written
    }

    private func failed() -> AutoBackupOutcome {
        store.update { $0.lastFailed = true }
        publish()
        return .failed
    }

    private func publish(destination: BackupDestination? = nil) {
        subject.send(AutoBackupState(record: store.record, destination: destination ?? subject.value.destination))
    }
}

/// The record in UserDefaults (the App Group suite the settings use), under Android's keys.
/// Unlike Android's, it rides along with iCloud Backup: the folder is the iCloud account's, so a
/// restored iPhone finds its copies where the record says.
final class UserDefaultsAutoBackupStore: AutoBackupStore {
    private let defaults: UserDefaults

    init(defaults: UserDefaults) {
        self.defaults = defaults
    }

    var record: AutoBackupRecord {
        AutoBackupRecord(
            enabled: defaults.object(forKey: Keys.enabled) as? Bool ?? true,
            lastBackupAt: (defaults.object(forKey: Keys.lastBackupAt) as? NSNumber)?.int64Value,
            lastFingerprint: defaults.string(forKey: Keys.lastFingerprint),
            lastFile: defaults.string(forKey: Keys.lastFile),
            lastFailed: defaults.bool(forKey: Keys.lastFailed),
            folderPromptDone: defaults.bool(forKey: Keys.folderPromptDone)
        )
    }

    func update(_ transform: (inout AutoBackupRecord) -> Void) {
        var value = record
        transform(&value)
        defaults.set(value.enabled, forKey: Keys.enabled)
        defaults.set(value.lastBackupAt.map { NSNumber(value: $0) }, forKey: Keys.lastBackupAt)
        defaults.set(value.lastFingerprint, forKey: Keys.lastFingerprint)
        defaults.set(value.lastFile, forKey: Keys.lastFile)
        defaults.set(value.lastFailed, forKey: Keys.lastFailed)
        defaults.set(value.folderPromptDone, forKey: Keys.folderPromptDone)
    }

    private enum Keys {
        static let enabled = "auto_backup_enabled"
        static let lastBackupAt = "auto_backup_last_backup_at"
        static let lastFingerprint = "auto_backup_last_fingerprint"
        static let lastFile = "auto_backup_last_file"
        static let lastFailed = "auto_backup_last_failed"
        static let folderPromptDone = "auto_backup_folder_prompt_done"
    }
}

/// An in-memory record, for tests and the UI-test container.
final class MemoryAutoBackupStore: AutoBackupStore {
    private(set) var record: AutoBackupRecord
    init(_ record: AutoBackupRecord = AutoBackupRecord()) { self.record = record }
    func update(_ transform: (inout AutoBackupRecord) -> Void) { transform(&record) }
}

/// The app's iCloud Drive folder: `Documents/` in its ubiquity container, which Files shows as
/// "Recipe Clipper" (`NSUbiquitousContainers` in Info.plist). Needs the iCloud Documents
/// capability and container on the App ID (docs/release.md); without them, or with iCloud Drive
/// off, the container is nil and the copy says it is unavailable.
final class ICloudBackupFolder: BackupFolder, @unchecked Sendable {
    static let containerIdentifier = "iCloud.com.liberopat.recipeclipper"

    private func directory() -> URL? {
        // Cheap first: nil while no iCloud account is signed in (or iCloud Drive is off).
        guard FileManager.default.ubiquityIdentityToken != nil,
              let container = FileManager.default.url(forUbiquityContainerIdentifier: Self.containerIdentifier)
        else { return nil }
        let documents = container.appendingPathComponent("Documents", isDirectory: true)
        try? FileManager.default.createDirectory(at: documents, withIntermediateDirectories: true)
        return documents
    }

    func destination() async -> BackupDestination {
        directory() == nil ? .unavailable : .ready
    }

    func list() async -> [String]? {
        guard let directory = directory(),
              let names = try? FileManager.default.contentsOfDirectory(atPath: directory.path) else { return nil }
        // A copy not downloaded to this iPhone is listed as ".<name>.icloud".
        return names.map { name in
            name.hasPrefix(".") && name.hasSuffix(".icloud") ? String(name.dropFirst().dropLast(7)) : name
        }
    }

    func write(name: String, json: String, photos: [String: URL]) async -> String? {
        guard let directory = directory() else { return nil }
        let data = BackupArchive.write(json: json, photos: photos.sorted { $0.key < $1.key }.map { (path: $0.key, file: $0.value) })
        var written = false
        var coordinationError: NSError?
        // Atomic: the file is replaced whole or not at all, so a half copy never stands.
        NSFileCoordinator().coordinate(
            writingItemAt: directory.appendingPathComponent(name), options: .forReplacing, error: &coordinationError
        ) { url in
            do {
                try data.write(to: url, options: .atomic)
                written = true
            } catch {
                dataLog.error("auto backup write failed: \(String(describing: error), privacy: .public)")
            }
        }
        return written && coordinationError == nil ? name : nil
    }

    func delete(name: String) async {
        guard let directory = directory() else { return }
        var coordinationError: NSError?
        NSFileCoordinator().coordinate(
            writingItemAt: directory.appendingPathComponent(name), options: .forDeleting, error: &coordinationError
        ) { url in
            try? FileManager.default.removeItem(at: url)
        }
    }
}

File: app/src/main/java/com/example/recipeclipper/data/AutoBackup.kt
package com.example.recipeclipper.data

import com.example.recipeclipper.data.backup.AutoBackupPolicy
import com.example.recipeclipper.data.backup.AutoBackupRecord
import com.example.recipeclipper.data.backup.BackupDestination
import com.example.recipeclipper.data.backup.BackupResult
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import javax.inject.Inject
import javax.inject.Singleton

/**
 * The folder the automatic copy (#150) goes to: a tree the user picked once through the system
 * picker, reached only through its lasting permission. The one part that needs a `Context`.
 */
interface BackupFolder {
    /** [BackupDestination.NOT_CHOSEN] with no folder, READY while its permission stands, else LOST. */
    fun destination(folderUri: String?): BackupDestination

    /** Keeps a lasting read-and-write permission on a picked folder; its name, or null if it can't. */
    fun adopt(folderUri: String): String?

    /** Gives a folder's permission back (the user chose another). */
    fun release(folderUri: String)

    /** The names of the files in the folder, or null if it can't be read. */
    suspend fun list(folderUri: String): List<String>?

    /** Writes a zip export ([com.example.recipeclipper.data.backup.BackupArchive]) as [name],
     *  replacing a file of that name; the name it ended up with, or null if it couldn't. A
     *  failed write leaves no partial file. */
    suspend fun write(folderUri: String, name: String, json: String, photos: Map<String, String>): String?

    /** Deletes the named file, if it is there. */
    suspend fun delete(folderUri: String, name: String)
}

/** Where [AutoBackupRecord] is kept: its own preferences file, outside Android's cloud backup. */
interface AutoBackupStore {
    val record: StateFlow<AutoBackupRecord>
    fun update(transform: (AutoBackupRecord) -> AutoBackupRecord)
}

/** When the app looks for a copy to write; WorkManager in the app, a recorder in tests. */
interface AutoBackupScheduler {
    /** A daily look, kept across restarts; idempotent. */
    fun startPeriodic()

    /** A look shortly after the app is left, coalescing quick returns. */
    fun afterLeaving()

    /** A look straight away (a folder was just chosen, or the copy turned back on). */
    fun now()
}

/** What Settings and Home show: the record, and whether its destination can be written now. */
data class AutoBackupState(val record: AutoBackupRecord, val destination: BackupDestination)

enum class AutoBackupOutcome { WRITTEN, UP_TO_DATE, OFF, NO_DESTINATION, FAILED }

/**
 * The automatic copy (#150): the latest export, photos included, written as a zip into the
 * folder the user chose, keeping the newest [AutoBackupPolicy.KEEP]. The rules are
 * [AutoBackupPolicy]'s; this runs them against the real export, folder and record. Nothing is
 * ever written anywhere else, and nothing is sent to any server.
 */
@Singleton
class AutoBackup @Inject constructor(
    private val backups: BackupRepository,
    private val folder: BackupFolder,
    private val store: AutoBackupStore,
    private val clock: Clock,
    private val scheduler: AutoBackupScheduler
) {
    // One copy at a time: the worker and "Back up now" may overlap.
    private val mutex = Mutex()

    // Bumped to recheck the folder's permission, which can go away while the app is closed.
    private val recheck = MutableStateFlow(0)

    val state: Flow<AutoBackupState> = combine(store.record, recheck) { record, _ ->
        AutoBackupState(record, folder.destination(record.folderUri))
    }

    val current: AutoBackupState
        get() = store.record.value.let { AutoBackupState(it, folder.destination(it.folderUri)) }

    /** Asks the folder again, e.g. when Settings opens. */
    fun recheckDestination() {
        recheck.value++
    }

    /** The daily look; called once per process. */
    fun start() = scheduler.startPeriodic()

    /** The app was left: a look soon, if a copy is being kept. */
    fun onAppLeft() {
        val record = store.record.value
        if (record.enabled && record.folderUri != null) scheduler.afterLeaving()
    }

    fun setEnabled(enabled: Boolean) {
        store.update { it.copy(enabled = enabled) }
        if (enabled) scheduler.now()
    }

    /** Home's one-time card was dismissed ("Not now"); Settings still offers the folder. */
    fun dismissFolderPrompt() = store.update { it.copy(folderPromptDone = true) }

    /**
     * The user picked a folder: keep its permission, give the old one back, and write a first
     * copy there at once. False if the permission couldn't be kept.
     */
    fun chooseFolder(folderUri: String): Boolean {
        val name = folder.adopt(folderUri) ?: return false
        val old = store.record.value.folderUri
        if (old != null && old != folderUri) folder.release(old)
        store.update {
            it.copy(enabled = true, folderUri = folderUri, folderName = name, folderPromptDone = true, lastFailed = false)
        }
        recheckDestination()
        scheduler.now()
        return true
    }

    /**
     * Writes a copy if [AutoBackupPolicy.isDue] (always, for [force]: "Back up now", which works
     * even with the automatic copy off). Old copies beyond the newest three go afterwards, so a
     * failed write never leaves fewer.
     */
    suspend fun run(force: Boolean = false): AutoBackupOutcome = mutex.withLock {
        val record = store.record.value
        if (!force && !record.enabled) return AutoBackupOutcome.OFF
        val folderUri = record.folderUri
        if (folderUri == null || folder.destination(folderUri) != BackupDestination.READY) {
            recheckDestination()
            return AutoBackupOutcome.NO_DESTINATION
        }
        val existing = folder.list(folderUri) ?: return failed()
        val exported = when (val result = backups.export()) {
            is BackupResult.Success -> result.value
            is BackupResult.Failure -> return failed()
        }
        val now = clock.now()
        val fingerprint = AutoBackupPolicy.fingerprint(exported.json, exported.photos.keys)
        if (!AutoBackupPolicy.isDue(record, fingerprint, existing, now, force)) return AutoBackupOutcome.UP_TO_DATE
        val written = folder.write(folderUri, AutoBackupPolicy.fileName(now), exported.json, exported.photos)
            ?: return failed()
        AutoBackupPolicy.toDelete(existing + written, written).forEach { folder.delete(folderUri, it) }
        store.update {
            it.copy(lastBackupAt = now, lastFingerprint = fingerprint, lastFile = written, lastFailed = false)
        }
        AutoBackupOutcome.WRITTEN
    }

    private fun failed(): AutoBackupOutcome {
        store.update { it.copy(lastFailed = true) }
        recheckDestination()
        return AutoBackupOutcome.FAILED
    }
}
